optimade/models/jsonapi.py

```diff
@@ -55,16 +56,39 @@ class JsonApi(BaseModel):
 class ToplevelLinks(BaseModel):
     """A set of Links objects, possibly including pagination"""
 
+    class Config:
+        extra = "allow"
+
     self: Optional[Union[AnyUrl, Link]] = Field(None, description="A link to itself")
     related: Optional[Union[AnyUrl, Link]] = Field(
         None, description="A related resource link"
     )
 
     # Pagination
-    first: Optional[AnyUrl] = Field(None, description="The first page of data")
-    last: Optional[AnyUrl] = Field(None, description="The last page of data")
-    prev: Optional[AnyUrl] = Field(None, description="The previous page of data")
-    next: Optional[AnyUrl] = Field(None, description="The next page of data")
+    first: Optional[Union[AnyUrl, Link]] = Field(
+        None, description="The first page of data"
+    )
+    last: Optional[Union[AnyUrl, Link]] = Field(
+        None, description="The last page of data"
+    )
+    prev: Optional[Union[AnyUrl, Link]] = Field(
+        None, description="The previous page of data"
+    )
+    next: Optional[Union[AnyUrl, Link]] = Field(
+        None, description="The next page of data"
+    )
+
+    @root_validator(pre=False)
+    def check_additional_keys_are_links(cls, values):
+        """The `ToplevelLinks` class allows any additional keys, as long as
+        they are also Links or Urls themselves.
+
+        """
+        for key, value in values.items():
+            if key not in cls.schema()["properties"]:
+                values[key] = parse_obj_as(Union[AnyUrl, Link], value)
```

Comment:
We can argue about whether this should also allow `Optional` (as below), but I think if an implementation is providing some arbitrary link with no value, its probably a mistake and they would rather know...

```suggestion
                values[key] = parse_obj_as(Optional[Union[AnyUrl, Link]], value)
```

Reply:
Njah. I don't know about that. I indeed think it should someone would do that. And it should definitely be allowed.

Reply:
Hmm, I have strong opposing feelings but no conviction so I'll roll over and commit the suggestion :upside_down_face: